runner/ollamarunner/runner.go

```diff
@@ -432,14 +431,21 @@ func (s *Server) processBatch() error {
 		}
 
 		// sample a token
-		vocabSize := len(f32s) / len(options.Outputs)
-		tokens, err := sample.Sample(logits[seq.iBatch*vocabSize:(seq.iBatch+1)*vocabSize], seq.samplers...)
+		vocabSize := len(logits) / len(options.Outputs)
+
+		var sampler sample.Sampler
+		if seq.useGreedy {
+			sampler = sample.Greedy()
+		} else {
+			sampler = sample.Weighted(nil)
+		}
+
+		sampledToken, err := sampler.Sample(logits[seq.iBatch*vocabSize:(seq.iBatch+1)*vocabSize], seq.transforms...)
```

Comment:
This seems unnecessary

```suggestion
		seqLogits = logits[seq.iBatch*vocabSize:(seq.iBatch+1)*vocabSize]
		var token int32
		if req.Temperature > 0 {
			token, err = sample.Weighted(seqLogits, seq.samplers)
		} else {
			token, err = sample.Greedy(seqLogits)
		}
```

where `sample.Greedy` and `sample.Weighted` are simply functions

Reply:
Good call - changed!

Reply:
I would rather store a sample.Sampler than spreading the sampling logic around everywhere, especially in `processBatch`, which is already too long. There's also no reason to recreate the sampling chain on every token. It's better to keep all of the sampling logic in `getTransforms`, though I would change the name back to something more general.

Besides readability, there is a more fundamental problem of what happens when you have stateful sampling? I think this is already a problem with the seed - the PRNG should only be seeded at the beginning of the sequence, not for every token.

Reply:
I think there's two ways to go about it based on chats:
1. Sampler package stays extremely simple and the user of the package can handle the logic to figure out what things to run how
2. Sampler package takes all sampling parameters as a pass through and manages it there

As for things like seed - I was just looking into that. We might need to change some of the runner logic and how we're doing batching? There needs to be a way for us to map each sequence to its own sampler. 

The user can also change sampling parameters on a request level which means at minimum we need to create a new sampler on change aside from the sequence to sampler mapping.